### PR TITLE
fix(ivy): let ngcc replace all switchable declarations for ESM formats

### PR DESCRIPTION
**Pull request by JoostK**
For the ESM2015/ESM5 formats, PRE_NGCC declarations may exist in files
that would not be considered by FileParser as that is specifically
designed to find files that contain classes with Angular decorators to
transform. Hence, such PRE_NGCC usages would not be rewritten.

This commit moves everything from Analyzer to DecoratorAnalyzer to
perform the existing decorator analysis, and introduces a separate
analysis for PRE_NGCC usages. Both analysis results are merged together
into a single AnalyzedFile object per source file, such that a Renderer
can still apply all analysis results at once per source file.

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[x] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] angular.io application / infrastructure changes
[ ] Other... Please describe:
```

## What is the current behavior?

ngcc does not replace all PRE_NGCC usages for ESM formats, for instance `ivy_switch_legacy.ts` would not have been converted.

## What is the new behavior?
The analysis for switchable declarations has been separated from the analysis of decorated classes, such that PRE_NGCC usages are rewritten even in files that do not contain decorated classes.

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

## Other information

@petebacondarwin PTAL

### CODE REVIEW

File: packages/compiler-cli/src/ngcc/src/packages/transformer.ts [hunk 1]
```diff
@@ -24,10 +25,13 @@ import {FileParser} from '../parsing/file_parser';
 import {Esm2015Renderer} from '../rendering/esm2015_renderer';
 import {Esm5Renderer} from '../rendering/esm5_renderer';
 import {FileInfo, Renderer} from '../rendering/renderer';
+import {isDefined} from '../utils';
+
 import {checkMarkerFile, writeMarkerFile} from './build_marker';
 import {EntryPoint, EntryPointFormat} from './entry_point';
 
 
+
```
Comment (petebacondarwin) [inline packages/compiler-cli/src/ngcc/src/packages/transformer.ts]: unnecessary?

File: packages/compiler-cli/src/ngcc/src/analyzer.ts [hunk 2]
```diff
@@ -5,104 +5,67 @@
...
-}
 
-function isMatchingHandler<A, M>(handler: Partial<MatchingHandler<A, M>>):
-    handler is MatchingHandler<A, M> {
-  return !!handler.match;
+  private findSwitchableFiles(): SwitchableFile[] {
+    return this.program.getSourceFiles()
+        .map(sourceFile => this.analyzeFileForSwitches(sourceFile))
+        .filter(isDefined);
+  }
```
Comment (petebacondarwin) [inline packages/compiler-cli/src/ngcc/src/analyzer.ts]: This feels like something that should be in the `FileParser` not `Analyzer`.

**Review by petebacondarwin — CHANGES_REQUESTED** [review]
Hi @JoostK 
Thanks for spotting this!
I would like to discuss the approach to architecting this into ngcc. The current PR seems to mix up the responsibilities somewhat. Although looking back at the design, the way I shoe-horned the NGCC switch stuff into the code is not so pretty either.

I think that we could come up with a way to keep the different parts focussed and decoupled better.

**Comment by petebacondarwin** [comment]
I think the architectural mistake I made was to assume that we should only be interested in files that contain classes - i.e. the `FileParser` classes only return such files, which then impacts throughout the rest of the code.

**Comment by petebacondarwin** [comment]
Perhaps we could make the `FileParser` pluggable? In this way, it would find all files that match some plugged-in criteria. For the most part that would be a file that contains a decorated class, but another plugin could match files that contain the `PRE_NGCC...` switch.

**Comment by petebacondarwin** [comment]
Then the `Analyzer` or `ClassAnalyzer` as you renamed it, would only care about `ParsedFile`s that contain classes, while a separate `SwitchableAnalyzer` would only care about files that contain `PRE_NGCC` markers.

**Comment by petebacondarwin** [comment]
And finally, the Renderer could be pluggable too, applying the various transformations from the Analysis results to the source code and writing it out.

**Comment by petebacondarwin** [comment]
Actually, thinking about it more, we should get rid of the `FileParser` interface altogether - moving that functionality into the `NgccReflectionHost` interface.

And while we are at it, we should see if we could create a design that could accommodate the DTS transformation in a more general way...

**Comment by JoostK** [comment]
@petebacondarwin thanks for your thorough review. Yes, I agree with your points. For now I didn't want to overhaul all of ngcc and have to touch all tests as a consequence, so that's why have stuck with most of the current structure.

I think the biggest elephant in the room is indeed the `FileParser`. It is tied too much to what the `Analyzer` (now `DecoratorAnalyzer`) is trying to do. While it is okay to separate the concerns of finding the files to analyze and performing the analysis itself, the current project structure does not reflect this well.

We need some sort of game plan to arrive at the proposed architecture. I don't think it's feasible, nor necessary, to do it all in a single PR.

**Comment by petebacondarwin** [comment]
I have a branch with some refactorings. I'll push it shortly, then we could discuss how that might be a better platform for this fix?

**Comment by petebacondarwin** [comment]
@JoostK - take a look at this PR as a potential strategy for this refactoring: https://github.com/angular/angular/pull/26082

**Pull request closed without merging**